spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala

```diff
@@ -980,3 +983,241 @@ trait DeltaTypeWideningTableFeatureTests extends BeforeAndAfterEach {
     }
   }
 }
+
+trait DeltaTypeWideningCheckConstraintsTests {
+  self: QueryTest with SharedSparkSession =>
+
+  test("check constraint with type change") {
+    withTable("t") {
+      sql("CREATE TABLE t (a byte) USING DELTA")
+      sql("ALTER TABLE t ADD CONSTRAINT ck CHECK (hash(a) > 0)")
+      sql("INSERT INTO t VALUES (2)")
+      checkAnswer(sql("SELECT hash(a) FROM t"), Row(1765031574))
+
+      // Changing the type of a column that a CHECK constraint depends on is not allowed.
+      checkError(
+        exception = intercept[DeltaAnalysisException] {
+          sql("ALTER TABLE t CHANGE COLUMN a TYPE SMALLINT")
+        },
+        errorClass = "DELTA_CONSTRAINT_DEPENDENT_COLUMN_CHANGE",
+        parameters = Map(
+          "operation" -> "change",
+          "columnName" -> "a",
+          "constraints" -> "delta.constraints.ck -> hash ( a ) > 0"
+      ))
+    }
+  }
+
+  test("check constraint on nested field with type change") {
+    withTable("t") {
+      sql("CREATE TABLE t (a struct<x: byte, y: byte>) USING DELTA")
+      sql("ALTER TABLE t ADD CONSTRAINT ck CHECK (hash(a.x) > 0)")
+      sql("INSERT INTO t (a) VALUES (named_struct('x', 2, 'y', 3))")
+      checkAnswer(sql("SELECT hash(a.x) FROM t"), Row(1765031574))
+
+      checkError(
+        exception = intercept[DeltaAnalysisException] {
+          sql("ALTER TABLE t CHANGE COLUMN a.x TYPE SMALLINT")
+        },
+        errorClass = "DELTA_CONSTRAINT_DEPENDENT_COLUMN_CHANGE",
+        parameters = Map(
+          "operation" -> "change",
+          "columnName" -> "a.x",
+          "constraints" -> "delta.constraints.ck -> hash ( a . x ) > 0"
+      ))
+
+      // Changing the type of a.y is allowed since it's not referenced by the CHECK constraint.
+      sql("ALTER TABLE t CHANGE COLUMN a.y TYPE SMALLINT")
+      checkAnswer(sql("SELECT * FROM t"), Row(Row(2, 3)))
+    }
+  }
+
+  test(s"check constraint with type evolution") {
+    withTable("t") {
+      sql(s"CREATE TABLE t (a byte) USING DELTA")
+      sql("ALTER TABLE t ADD CONSTRAINT ck CHECK (hash(a) > 0)")
+      sql("INSERT INTO t VALUES (2)")
```

Comment:
Let's use a number above 127 to clearly show it will not be a TINYINT

Reply:
Now using larger values after the type has been widened